src/main/java/com/amazon/ion/impl/IonCursorBinary.java

```diff
@@ -3104,6 +3182,317 @@ public Event fillArgumentEncodingBitmap(int numberOfBytes) {
+            if (continuation == 0) {
+                // End of the group
+                event = Event.NEEDS_INSTRUCTION;
+                return event;
+            }
+            group.pageEndIndex = peekIndex + continuation;
+        }
+        if (group.pageEndIndex > limit && fillArgumentGroupPage(group)) {
+            return event;
+        }
+        // TODO performance: for fixed-width tagless types, the following could be skipped after the first value.
+        nextTaglessValue(group.primitiveType);
+        return event;
+    }
+
+    /**
+     * Positions the cursor on the next value in the group. Upon return, the value will be filled and `valueMarker` set
+     * to the value's start and end indices. This method may return:
+     * <ul>
+     *     <li>NEEDS_DATA, if not enough data is available in the stream</li>
+     *     <li>START_SCALAR, if the reader is now positioned on a scalar value</li>
+     *     <li>START_CONTAINER, if the reader is now positioned on a container value</li>
+     *     <li>NEEDS_INSTRUCTION, if the cursor reached the end of the argument group. Subsequently, the caller must
+     *     call {@link #exitArgumentGroup()}.</li>
+     * </ul>
+     * @return an Event conveying the result of the operation.
+     */
+    public Event nextGroupedValue() {
+        ArgumentGroupMarker group = argumentGroupStack[argumentGroupIndex];
+        if (peekIndex < valueMarker.endIndex) {
+            peekIndex = valueMarker.endIndex;
+        }
+        if (group.primitiveType == null) {
+            return nextGroupedTaggedValue(group);
+        }
+        return nextGroupedTaglessValue(group);
+    }
+
+    /**
+     * Seeks the cursor to the end of the current page of the argument group.
+     * @param group the group in which to seek.
+     * @return true if there was not enough data to complete the seek; otherwise, false.
+     */
+    private boolean seekToEndOfArgumentGroupPage(ArgumentGroupMarker group) {
+        if (isSlowMode) {
+            if (slowSeek(group.pageEndIndex - offset)) {
+                return true;
+            }
+            peekIndex = offset;
+        } else {
+            peekIndex = group.pageEndIndex;
+        }
+        return false;
+    }
+
+    // Dummy delimited container to be used when seeking forward to a delimited end marker of a synthetic container,
+    // like an argument group.
+    private static final IonTypeID DUMMY_DELIMITED_CONTAINER = TYPE_IDS_1_1[OpCodes.DELIMITED_SEXP & SINGLE_BYTE_MASK];
+
+    /**
+     * Seeks to the end of the current delimited argument group.
+     * @return true if not enough data was available to complete the seek; otherwise, false.
+     */
+    private boolean seekToEndOfDelimitedArgumentGroup() {
+        // Push a dummy delimited container onto the stack, preparing the cursor to seek forward to the delimited end
+        // marker applicable at the current depth.
+        pushContainer();
```

Comment:
What is your reasoning for not actually treating expression groups like a container?

If you just treated expression groups like another container, it seems like you could probably eliminate some data structures, some code that is similar between containers and expression groups, and some public methods.

Reply:
I considered implementing it that way, but decided not to (for now) for a few reasons:
1. It complicates calculating of the reader's current depth because the argument groups need to be filtered out. This is pretty minor.
2. The container stack uses the Marker class, which is reused in many places and has a startIndex, endIndex, and IonTypeID. Argument groups need a startIndex, endIndex, and PrimitiveType. I'd have to merge these classes if using a unified stack, and I didn't feel like bloating Marker with an attribute that is only needed by argument groups (and is never needed in Ion 1.0). This also isn't a huge deal.

Which public methods were you envisioning this would allow us to remove? I'm not sure we could reuse the stepIn/stepOut methods for this, for example, because the encoding for tagged and tagless groups differs and is determined by the signature.

Anyway, I'm happy to revisit this in the future, but if you're up for merging this as-is for now we can start using it and see whether it ends up being cumbersome.